Add fallback for non-primary site primary section.

Shinsina opened the pull request:
This will use the first section on the given site within the content items schedules and set it as the primary section.

Example article is one from Food Logistics on SDCE (https://www.foodlogistics.com/professional-development/training/news/22031062/gartner-inc-supply-chain-leaders-expect-permanent-hybrid-work-model-for-frontline-workers)

<img width="1920" alt="Screen Shot 2022-02-03 at 2 28 55 PM" src="https://user-images.githubusercontent.com/46794001/152423765-ba52969f-28c2-4beb-814e-0cc8b88f267e.png">

<img width="1920" alt="Screen Shot 2022-02-03 at 2 28 47 PM" src="https://user-images.githubusercontent.com/46794001/152423752-6b541900-cf93-4403-b4df-26e91ef910dd.png">
.

Shinsina committed:
Add fallback for non-primary site primary section.

This will use the first section on the given site within the content items schedules and set it as the primary section.

Shinsina commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Content primary to food is showing to SDC as 'home'](https://trello.com/c/VweV2Fi2/167-content-primary-to-food-is-showing-to-sdc-as-home)

Shinsina commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Why does "Home" show on some content bars.](https://trello.com/c/dRB9Y2tC/227-why-does-home-show-on-some-content-bars)

solocommand reviewed (changes requested):
@Shinsina This is great! One change -- use the `needs` parameter to the `@projection` directive on the definition and you won't need to do another `load`/query of the content.
![image](https://user-images.githubusercontent.com/1778222/152424766-3fc51218-0896-4367-b125-3a5fa8aac467.png)
In your resolver, just use `getAsArray` to retrieve the `sectionQuery` field from the `content` item.

Shinsina committed:
Add sectionQuery to projection directive.

Shinsina committed:
Use getAsArray for sectionQuery off content.

zarathustra323 commented on services/graphql-server/src/graphql/resolvers/platform/content.js:
```diff
@@ -447,7 +447,7 @@ module.exports = {
      * Load primary section of content.
      * If primary section's site matches the current site, return the section.
-     * If not, check for alternative site + section (@todo).
+     * If not, check for alternative site + section.
```
The todo comments should actually stay, as they are referring to explicitly allowing a user to set a secondary site/section, not a fallback to a schedule

zarathustra323 commented on services/graphql-server/src/graphql/resolvers/platform/content.js:
```diff
@@ -482,7 +482,17 @@ module.exports = {
 
       // Current section does not match site, load alternate.
-      // @todo This should eventually account for secondary sites/sections. For now, load home.
+      const sectionQuery = getAsArray(content, 'sectionQuery');
```
This can be accomplished by returning _one_ section, not all of them. No need to query all possible sections when you're only returning one of them. :)

Instead, change your query to include the current siteId context, include where the schedule section Ids exist, and where the alias _is not_ home

solocommand replied:
> This can be accomplished by returning _one_ section, not all of them. No need to query all possible sections when you're only returning one of them. :)
> 
> Instead, change your query to exclude the current siteId context, include where the schedule section Ids exist, and where the alias _is not_ home

The current siteId context is correct at this point (it _should_ be included, not excluded), but agreed on `findOne` instead of `find` and the alias

zarathustra323 replied:
Yep, already fixed the siteId part

zarathustra323 replied:
```js
const currentSiteSectionQuery = {
  ...formatStatus(status),
  'site.$id': siteId,
  _id: { $in: sectionQuery.map(schedule => schedule.sectionId) },
  alias: { $ne: 'home' },
};
```

solocommand replied:
@Shinsina you can copy the website context exception from the `siteContext` field to ensure that a site id is present.

zarathustra323 commented:
@solocommand and @Shinsina Ideally the `siteId` should be saved on each subdocument of the `Content.sectionQuery` array... if it were present, the resolver could filter the array ahead of time and load the section via a dataloader using the section ID (instead of doing a hard query). The benefit is that the dataloader would then be able to batch multiple requests... so a list of 50 content items would result in one query to the section collection, where right now this logic would result in 50 _individual_ new queries to get the fallback section (assuming all needed fallback)

Shinsina committed:
Return one section,bring back todos,verify siteId.

Shinsina commented:
@solocommand and @zarathustra323 Additionally we'll need to consider doing the same thing here: https://github.com/parameter1/base-cms/blob/master/packages/canonical-path/src/platform/content.js#L21 so it would likely make sense to script something to add siteId to each subdocument of ```Content.sectionQuery``` (for already existing content) and modify the interface to additionally account for adding/handling that field when schedules are created. 

I think I managed to make sense of the previous comments in my most recent commit but if I missed something let me know.

zarathustra323 commented:
@Shinsina yes, good point. The canonical path would also need to account for this... so... adjusting the data is probably necessary. Though such a change touches a lot of places. @solocommand thoughts?

Shinsina committed:
Fix return logic rename to currentSiteSection.

Shinsina committed:
Filter sectionQuery and use dataloader over query.

Shinsina committed:
Use identical logic for path resolution.

Shinsina committed:
Return initial section check if found.

Shinsina committed:
Return section alias over section object.

Shinsina commented:
@zarathustra323 @solocommand I've corrected the resolution logic for primarySection to utilize siteId off the sectionQuery subdocuments and used a dataloader to grab the appropriate section, additionally I added (essentially) identical logic for the path resolution as well. 

Let me know if there is anything I missed or additional corrections that need to be made.

Shinsina committed:
Remove unnecessary logical AND, sectionQuery check

solocommand commented on packages/canonical-path/src/platform/content.js:
```diff
@@ -18,9 +18,22 @@ const pathResolvers = {
+        const currentSiteSection = await load('websiteSection', currentSiteSections[0].sectionId, { alias: 1 }, query);
+        if (currentSiteSection) return currentSiteSection.alias;
+      }
+    }
```
Use `Array.find` instead of map+filter when returning a single result.
```suggestion
    const item = sectionQuery.find(({ siteId }) => siteId === site.id());
    if (item) {
      const secondary = await load('websiteSection', item.sectionId, { alias: 1 }, query);
      if (secondary) return secondary.alias;
    }
```

solocommand commented on services/graphql-server/src/graphql/resolvers/platform/content.js:
```diff
@@ -482,7 +482,20 @@ module.exports = {
+          const currentSiteSection = await load('websiteSection', currentSiteSections[0].sectionId, projection, query);
+          if (currentSiteSection) return currentSiteSection;
+        }
+      }
```
Same change here. Since you're doing the same thing in both files, export this logic as a utility function. It should take the `content` object and `load` function as parameters, and return the found section object (if present).

In the path resolver, you'd just return the alias from the secondary section, if it was found.

Shinsina committed:
Create utility function.

Shinsina committed:
Use utility function in both previous locations.

Shinsina committed:
Correct projection in load function.

Shinsina committed:
Use correct pathing for require statements

Shinsina committed:
Remove added space.

Shinsina committed:
Do not return home.

Shinsina committed:
Reduce filter logical complexity.

Shinsina committed:
Spread Promises into Promise.all.

Shinsina committed:
Return Promise array directly.

Shinsina committed:
Modify query logic to not return home.

Shinsina committed:
.length check for returning found section

Shinsina committed:
Return a working result.

zarathustra323 commented on packages/utils/src/get-section-from-schedules.js:
```diff
@@ -0,0 +1,15 @@
+module.exports = async ({ content, siteId, load }) => {
+  const { sectionQuery } = content;
```
Add `if (!sectionQuery.length) return null;` as the first line, then delete the `if` wrapper on L8 and the other `return null` on L12

zarathustra323 commented on packages/utils/src/get-section-from-schedules.js:
```diff
@@ -0,0 +1,15 @@
+    alias: { $ne: 'home' },
+  };
+  if (sectionQuery.length) {
+    const sectionsFromScheds = sectionQuery.filter(schedule => schedule.siteId === siteId);
```
These need to be stringify if they're MongoIds otherwise they'll never match.

`ObjectId()` does not equal `ObjectId()` because they are different object instances....

Change to
```js
`${schedule.siteId}` === `${siteId}`
```

zarathustra323 commented on packages/utils/src/get-section-from-schedules.js:
```diff
@@ -0,0 +1,15 @@
+  };
+  if (sectionQuery.length) {
+    const sectionsFromScheds = sectionQuery.filter(schedule => schedule.siteId === siteId);
+    const foundSections = await Promise.all(sectionsFromScheds.map(section => load('websiteSection', section.sectionId, {}, query)));
```
The `projection` argument should not be an empty object, otherwise all fields will return. Add the `projection` argument to this util function and pass it to the `load` function

Shinsina replied:
I need all the fields as this is also being used to return the section itself here: https://github.com/parameter1/base-cms/pull/233/files/bf814341b5cb4a668917cb74833ab34425de5490#diff-466f539f3ec0e3fd7f2308c36f524ccf2ff92b8ef482f01cf0cf5d0eeb98e4faR488

zarathustra323 replied:
> Do we want it to just return the first section regardless of whether or not it is aliased as "home",

@Shinsina yes, you are correct... I forgot about the alias requirement. I'll adjust my comments where necessary

zarathustra323 replied:
@Shinsina the projection should be passed in by the callees, (see my comment on L487 of `resolvers/platform/content.js`)

zarathustra323 replied:
Also, I removed the comments asking to change the `filter` to `find`

zarathustra323 commented on services/graphql-server/src/graphql/resolvers/platform/content.js:
```diff
@@ -482,7 +482,10 @@ module.exports = {
-      // @todo This should eventually account for secondary sites/sections. For now, load home.
+      // @todo This should eventually account for secondary sites/sections.
+      // For now load an alternate from schedules.
+      const sectionFromSched = await getSectionFromSchedules({ content, siteId, load });
```
For clarity, might as well name the variable the full name ... `sectionFromSchedule`

zarathustra323 replied:
Pass `projection` through as an argument parameter

zarathustra323 commented on packages/canonical-path/src/platform/content.js:
```diff
@@ -18,9 +18,13 @@ const pathResolvers = {
-    // For now load home when not found.
-    return section ? section.alias : 'home';
+    // For now load an alternate from schedules
+    const sectionFromSched = await getSectionFromSchedules({ content, siteId: site.id(), load });
```
The projection here should be explicitly passed as `{ alias: 1 }`

Shinsina commented:
@zarathustra323 Do we want it to just return the first section regardless of whether or not it is aliased as "home", say I have an article scheduled to home, a section called foo and a section called bar on the non-primary site, if the sectionQuery just so happens to have it in the order of Home > foo > bar should it return home or a "more specific" section?

Shinsina committed:
Pass in projection check siteId with matching type

Shinsina committed:
Rename assigned var, pass in projection.

Shinsina committed:
Use load over Promise.all mapping.

I believe if I read the README correctly load will accept an array as it's "query" and spread each one out into it's own Promise.all anyway, this should make this less "logically complicated" in that we don't have to query for each one individually.

Shinsina committed:
Revert previous.

Shinsina committed:
Filter correctly.
